akka-typed/src/main/scala/akka/typed/ScalaProcess.scala

```diff
@@ -0,0 +1,538 @@
+    private def unit[S, Out]: Operation[S, Out] = _unit.asInstanceOf[Operation[S, Out]]
+
+    def loopInf[S]: NextLoopInf[S] = nextLoopInf.asInstanceOf[NextLoopInf[S]]
+    trait NextLoopInf[S] {
+      def apply[U](body: OpDSL { type Self = S } ⇒ Operation[S, U]): Operation[S, Nothing] = {
+        lazy val l: Operation[S, Nothing] = unit[S, OpDSL { type Self = S }].flatMap(body).flatMap(_ ⇒ l)
+        l
+      }
+    }
+    private object nextLoopInf extends NextLoopInf[Nothing]
+
+    def loop[S]: NextLoop[S] = nextLoop.asInstanceOf[NextLoop[S]]
+    trait NextLoop[S] {
+      def apply[Out](n: Int)(body: OpDSL { type Self = S } ⇒ Operation[S, Out]): Operation[S, List[Out]] = {
+        require(n > 0, "number of iterations must be positive")
+        def step(n: Int, acc: List[Out]): Operation[S, List[Out]] =
+          unit[S, OpDSL { type Self = S }]
+            .flatMap(body)
+            .flatMap {
+              case result if n == 1 ⇒ Return((result :: acc).reverse)
+              case result           ⇒ step(n - 1, result :: acc)
+            }
+        step(n, Nil)
+      }
+    }
+    private object nextLoop extends NextLoop[Nothing]
+
+    def apply[T]: Next[T] = next.asInstanceOf[Next[T]]
+
+    trait Next[T] {
+      def apply[U](body: OpDSL { type Self = T } ⇒ Operation[T, U]): Operation[T, U] =
+        unit[T, OpDSL { type Self = T }].flatMap(body)
+    }
+    private object next extends Next[Nothing]
+
+    trait NextStep[T] {
+      def apply[U](mailboxCapacity: Int, body: OpDSL { type Self = T } ⇒ Operation[T, U])(implicit opDSL: OpDSL): Operation[opDSL.Self, U] =
+        Call(Process("nextStep", Duration.Inf, mailboxCapacity, body(null)))
+    }
+    private[typed] object nextStep extends NextStep[Nothing]
+  }
+
+  /*
+   * Terminology:
+   *
+   *  - a Process has a 1:1 relationship with an ActorRef
+   *  - an Operation is a step that a Process takes and that produces a value
+   *  - Processes are concurrent, but not distributed: all failures stop the entire Actor
+   *  - each Process has its own identity (due to ActorRef), and the Actor has its own
+   *    identity (an ActorRef[ActorCmd[_]]); processSelf is the Process’ identity, actorSelf is the Actor’s
+   *  - timeout means failure
+   *  - every Actor has a KV store for state
+   *      - querying by key in TypedMultiMap (using a single element per slot)
+   *      - updating is an Operation taking an Event and an implicit (Event, State) => State
+   *      - persistence can then be plugged in transparently
+   *      - recovery means acquiring state initially (which might trigger internal replay)
+   */
+
+  /**
+   * Helper to make `Operation.map` behave like `flatMap` when needed.
+   */
+  trait MapAdapter[Self, Out, Mapped] {
+    def lift[O](f: O ⇒ Out): O ⇒ Operation[Self, Mapped]
+  }
+  /**
+   * Helper to make `Operation.map` behave like `flatMap` when needed.
+   */
+  object MapAdapter extends MapAdapterLow {
```

Comment:
Boring but let's keep the `...LowPriorityImplicits` naming (we have it in other places)